### PR TITLE
NetSim: Fix disconnect/reconnect bug

### PR DESCRIPTION
**Pull request by islemaster**
**The problem**
[Bug #95758778](https://www.pivotaltracker.com/story/show/95758778): NetSim silently fails to send a message, with the following reliable repro steps:
1. Load a NetSim level and connect to a section.
2. Connect to a router.
3. Send a message to yourself. It should be routed successfully.
4. Disconnect from the router and then reconnect to the same router (not a page reload!)
5. Send another message to yourself.  This time the message shows up in your sent log, but it will not show up in the router log or be echoed back to you.

**The root cause**
Thirteen days ago when I added inter-router messaging, I made [clients simulate all of the routers](https://github.com/code-dot-org/code-dot-org/commit/6c24e60dc8258b8e619fa3ef37839e656cf5f3f8#diff-f6f236824ac55c69828e680581f110bdR528), not just their direct connection.  When the client connected directly to a router they still created a second simulating instance of that router, which seemed messy.

The next day I [cleaned this up](https://github.com/code-dot-org/code-dot-org/commit/c69521671100a5bbd58fa9ea48cf57748c8bd66f#diff-f6f236824ac55c69828e680581f110bdL307), removing the extra instance and instead pointing it to the router in the existing collection.  In the process, I removed the call to `initializeSimulation` on connect (because the router was already initialized when it was first detected).

I [forgot to remove the `stopSimulation` calls on disconnect](https://github.com/code-dot-org/code-dot-org/commit/c69521671100a5bbd58fa9ea48cf57748c8bd66f#diff-f6f236824ac55c69828e680581f110bdR450).  So disconnecting from a router stops its simulation, but connecting to it doesn't restart the simulation - hence, messages aren't picked up by the router after doing a disconnect/reconnect.

**The solution**
Here, I remove the offending `stopSimulation` calls on disconnect - router simulation should only stop [when we detect that the router is no longer on the shard](https://github.com/code-dot-org/code-dot-org/commit/6c24e60dc8258b8e619fa3ef37839e656cf5f3f8#diff-f6f236824ac55c69828e680581f110bdR535).  I also balance the remaining item that's unique to the directly-connected router: We only provide a heartbeat for one router at a time, so we manually attach the heartbeat controller on connect and (as of this PR) manually detach it on disconnect.

I've added some extra diagnostic logging, but bumped the default log level up to 'Warn,' to prevent log noise in production.  We can set it to 'Verbose' as needed to diagnose issues.

### CODE REVIEW

**Comment by islemaster** [comment]
Update: Took me a bit, but I got a test written for this.

File: apps/test/netsim/NetSimRouterNode.js [hunk 1]
```diff
@@ -172,14 +187,42 @@ describe("NetSimRouterNode", function () {
...
+  };
+
+  /**
+   * Build a fake message log that can be passed into
+   * NetSimLocalClientNode.initializeSimulation as the sent or received log,
+   * and can be used to sense whether a particular message would be passed
+   * to the sent/received log UI components.
+   * @returns {{log: Function, getLatest: Function}}
+   */
+  var makeSensingMessageLog = function () {
```
Comment (davidsbailey) [inline apps/test/netsim/NetSimRouterNode.js]: There are well-defined names for various types of objects used for testing: http://en.wikipedia.org/wiki/Test_double

This looks like a fake, so let's call it makeFakeMessageLog.

File: apps/test/netsim/NetSimRouterNode.js [hunk 2]
```diff
@@ -744,6 +787,33 @@ describe("NetSimRouterNode", function () {
 
   });
 
+  it ("still simulates/routes after connect-disconnect-connect cycle", function () {
+    netsimGlobals.getLevelConfig().automaticReceive = true;
+    var time = 1;
+    var receivedLog = makeSensingMessageLog();
```
Comment (davidsbailey) [inline apps/test/netsim/NetSimRouterNode.js]: if we agree this is a fake, let's call it fakeReceivedLog.

**Comment by davidsbailey** [comment]
LGTM. There are quite a few methods around destroying and disconnecting now which took me a few minutes to understand. I don't see an obvious way to make it any simpler than you have already, except maybe a boolean sync/async parameter (arguable, optional, mostly thinking out loud).

File: apps/test/netsim/NetSimRouterNode.js (hunk 1 above)
Comment (islemaster) [inline apps/test/netsim/NetSimRouterNode.js]: That sounds good.  I was thinking of "Introduce sensing variable" from Feathers' _Legacy Code_ but you're right, this is a fake.  Updating...

**Comment by islemaster** [comment]
Yeah, disconnect stuff is a bit of a mess.  I went ahead and kept the synchronous stuff separate for now, in part because it should _only_ be used when responding to the onUnload event so it's a very special case.

**Pull request merged**